File: 2020/15/Main.scala
package main

import scala.io.Source
import scala.collection.mutable.ListBuffer

object  Main extends App {
  def next(in: List[Int]): Int = {
    val last = in.last


    if (in.count(_ == last) == 1) {
      // first time spoken
      return 0
    }

    // else find index

    return in.reverse.drop(1).takeWhile(_ != last).length + 1
  }

  def solve1(): Int = {
    var numbers = ListBuffer[Int](2,0,6,12,1,3)
    //var numbers = ListBuffer[Int](0,3,6)

    var iter = numbers.toList.length
    while (true) {
      val n = next(numbers.toList)
      numbers += n
      iter = iter + 1
      if (iter == 2020) {
        return n
      }

    }

    return 0
  }

  println(solve1())
}

File: 2020/15/main.go
package main

import (
	"fmt"
)

var m map[int][]int

func main() {
	m = map[int][]int{} // map of key -> indices
	fmt.Printf("%v\n", solve2())
}

func next(in []int) int {
	last := in[len(in)-1]

	if indices, ok := m[last]; !ok || len(indices) == 1 {
		return 0
	} else {
		i1, i2 := indices[len(indices)-1], indices[len(indices)-2]
		return i1 - i2
	}
}

func solve2() int {
	numbers := getInput()
	for i, n := range numbers {
		m[n] = append(m[n], i)
	}
	var n int
	for i := len(numbers); i != 30000000; i++ {
		n = next(numbers)
		numbers = append(numbers, n)
		m[n] = append(m[n], i)
	}
	return n
}

func getInput() []int {
	return []int{2, 0, 6, 12, 1, 3}
}
